cpp/src/reductions/struct_minmax_util.cuh

```diff
@@ -87,49 +83,39 @@ auto static constexpr DEFAULT_NULL_ORDER = cudf::null_order::BEFORE;
  */
 class comparison_binop_generator {
  private:
-  std::unique_ptr<cudf::structs::detail::flattened_table> const flattened_input;
-  std::unique_ptr<table_device_view, std::function<void(table_device_view*)>> const
-    d_flattened_input_ptr;
+  cudf::table_view const input_tview;
   bool const is_min_op;
   bool const has_nulls;
-
   std::vector<null_order> null_orders;
-  rmm::device_uvector<null_order> null_orders_dvec;
-
-  comparison_binop_generator(column_view const& input, rmm::cuda_stream_view stream, bool is_min_op)
-    : flattened_input{cudf::structs::detail::flatten_nested_columns(
-        table_view{{input}},
-        {},
-        std::vector<null_order>{DEFAULT_NULL_ORDER},
-        cudf::structs::detail::column_nullability::MATCH_INCOMING,
-        stream,
-        rmm::mr::get_current_device_resource())},
-      d_flattened_input_ptr{
-        table_device_view::create(flattened_input->flattened_columns(), stream)},
-      is_min_op(is_min_op),
-      has_nulls{has_nested_nulls(table_view{{input}})},
-      null_orders_dvec(0, stream)
+  rmm::cuda_stream_view stream;
+
+  comparison_binop_generator(column_view const& input_,
+                             rmm::cuda_stream_view stream_,
+                             bool is_min_op_)
+    : input_tview{cudf::table_view{{input_}}},
+      is_min_op(is_min_op_),
+      has_nulls{cudf::has_nested_nulls(cudf::table_view{{input_}})},
+      null_orders{std::vector<null_order>{DEFAULT_NULL_ORDER}},
```

Comment:
I think this is the reason that fails many tests in this PR. In the meantime I don't have a clear way to fix it. Need to think further.

Reply:
Only this test is failing in this PR now. Any ideas why?
```
StructReductionTest.StructReductionMinMaxWithNulls
```

Reply:
I'm not sure. Can you print out the output of that test to see what it is?

Reply:
```
[ RUN      ] StructReductionTest.StructReductionMinMaxWithNulls
/opt/conda/conda-bld/work/cpp/tests/utilities/column_utilities.cu:289: Failure
Expected equality of these values:
  lhs_null_count
    Which is: 1
  rhs_null_count
    Which is: 0
Google Test trace:
/opt/conda/conda-bld/work/cpp/tests/reductions/reduction_tests.cpp:2565:  <--  line of failure
```

Reply:
```

Reply:
It seems that the `min` test fails, and `aab` is considered as smallest, and nulls are excluded (nulls are larger than non-nulls).

Reply:
It looks like this was already a [pre-conditon](https://github.com/rapidsai/cudf/blob/aed7174eae6c6eb38fbf186938df44f88787cf29/cpp/src/reductions/struct_minmax_util.cuh#L113-L124) for `min` where top-level NULLs were pushed to `AFTER`

Reply:
Yes. So the correct output is `{null, null}` which contains nulls only in the children, not top level.
In other word: nulls are excluded (nulls are larger than non-nulls) in the children level.

Reply:
Ah I see, I don't understand why this is happening. Fault with code I wrote or lex comparator?

Reply:
The previous approach flattens the input structs column into a table. The null order here will have size equal to number of table columns:
```
  if (is_min_op) {
      null_orders = flattened_input->null_orders();
```
So we could just change the first null order element and the top level null order will be different from all children level null order.

In the new approach, you always have just one input column and always have the null order array having one element. The row comparator will then use that unique null order for all top level and children level.

 I can think of only one solution to overcome this: continue flattening the input and modify the null order vector, then pass everything into the row comparator.

Reply:
And that is only needed for `min`. For `max`, all null order values are the same.